### PR TITLE
refactor: EmailContentProvider classes to use java objects (FPLA-1513)

### PR DESCRIPTION
**Pull request by aliveni**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/FPLA-1513

### Change description ###
Created java objects for email template data which are using `Map`.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

**Review by lewisbirks — COMMENTED** [review]
Will be a bit annoying to do but I feel like all the notify models should just use `@Data` or `@Value` and the `@SuperBuilder`/ `@Builder`. Will result in a lot less noise at the top of the classes

**Comment by lewisbirks** [comment]
If jackson then ends up complaining about things use `@Jacksonized` to tell jackson to use the builders for serialisation

**Comment by aliveni** [comment]
> Will be a bit annoying to do but I feel like all the notify models should just use `@Data` or `@Value` and the `@SuperBuilder`/ `@Builder`. Will result in a lot less noise at the top of the classes

@lewisbirks added the suggested annotations to all notify models.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/notify/cmo/CMOReadyToSealTemplate.java [hunk 1]
```diff
@@ -1,13 +1,13 @@
 package uk.gov.hmcts.reform.fpl.model.notify.cmo;
 
-import lombok.Getter;
-import lombok.Setter;
+import lombok.Data;
 import lombok.experimental.Accessors;
+import lombok.experimental.SuperBuilder;
 import uk.gov.hmcts.reform.fpl.model.notify.NotifyData;
 
 @Accessors(chain = true)
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/notify/cmo/CMOReadyToSealTemplate.java:8]: Can remove this now that we are using the builder

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/notify/allocatedjudge/SharedAllocatedJudgeTemplate.java [hunk 2]
```diff
@@ -1,11 +1,13 @@
...
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import lombok.experimental.SuperBuilder;
 import uk.gov.hmcts.reform.fpl.model.notify.NotifyData;
 
-@Getter
-@Setter
+@Data
+@SuperBuilder
+@NoArgsConstructor
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/notify/allocatedjudge/SharedAllocatedJudgeTemplate.java:10]: Is there a reason for requiring this?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/notify/cmo/DraftCMOUploadedTemplate.java [hunk 3]
```diff
@@ -1,13 +1,13 @@
 package uk.gov.hmcts.reform.fpl.model.notify.cmo;
 
-import lombok.Getter;
-import lombok.Setter;
+import lombok.Data;
 import lombok.experimental.Accessors;
+import lombok.experimental.SuperBuilder;
 import uk.gov.hmcts.reform.fpl.model.notify.NotifyData;
 
 @Accessors(chain = true)
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/notify/cmo/DraftCMOUploadedTemplate.java:8]: Can be removed now that a builder is used

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/notify/cmo/DraftCMOUploadedTemplate.java [hunk 4]
```diff
@@ -1,13 +1,13 @@
...
-@Getter
-@Setter
+@Data
+@SuperBuilder
 public class DraftCMOUploadedTemplate implements NotifyData {
     private String judgeTitle;
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/notify/cmo/DraftCMOUploadedTemplate.java:12]: You have made things final elsewhere, might as well do the same here

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/notify/hearing/UpcomingHearingNotifyData.java [hunk 5]
```diff
@@ -0,0 +1,14 @@
+package uk.gov.hmcts.reform.fpl.model.notify.hearing;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import lombok.Builder;
+import lombok.EqualsAndHashCode;
+import uk.gov.hmcts.reform.fpl.model.notify.NotifyData;
+
+@Builder
+@EqualsAndHashCode
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/notify/hearing/UpcomingHearingNotifyData.java:9]: `@Data`?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/notify/sendtogatekeeper/NotifyGatekeeperTemplate.java [hunk 6]
```diff
@@ -1,28 +1,28 @@
...
-        clone.setTimeFrameValue(this.getTimeFrameValue());
-        clone.setUrgentHearing(this.getUrgentHearing());
-        clone.setNonUrgentHearing(this.getNonUrgentHearing());
-        clone.setFirstRespondentName(this.getFirstRespondentName());
-        clone.setReference(this.getReference());
-        clone.setCaseUrl(this.getCaseUrl());
-
-        return clone;
+    public SharedNotifyTemplate duplicate() {
+        return SharedNotifyTemplate.builder()
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/notify/sendtogatekeeper/NotifyGatekeeperTemplate.java:13]: Can just be `this.toBuilder().build()`

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/notify/submittedcase/SubmitCaseCafcassTemplate.java [hunk 7]
```diff
@@ -1,11 +1,13 @@
...
 import uk.gov.hmcts.reform.fpl.model.notify.SharedNotifyTemplate;
 
-@Getter
-@Setter
+@EqualsAndHashCode(callSuper = true)
+@SuperBuilder
+@Data
 public final class SubmitCaseCafcassTemplate extends SharedNotifyTemplate {
-    private String cafcass;
+    String cafcass;
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/notify/submittedcase/SubmitCaseCafcassTemplate.java:12]: is there a reason for this no longer being private?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/base/AbstractEmailContentProvider.java [hunk 8]
```diff
@@ -26,6 +27,10 @@ public String getCaseUrl(Long caseId) {
         return caseUrlService.getCaseUrl(caseId);
     }
 
+    public String getCaseUrl(CaseData caseData) {
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/base/AbstractEmailContentProvider.java:30]: Can't find where this is being used. Can we remove if not needed?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/base/StandardDirectionOrderContent.java [hunk 9]
```diff
@@ -1,33 +1,24 @@
...
-            .put("hearingDate", getHearingBooking(caseData))
-            .put("reference", String.valueOf(caseData.getId()))
-            .put("callout", buildCallout(caseData))
-            .put("caseUrl", getCaseUrl(caseData.getId(), "OrdersTab"));
+    protected String getFamilyManCaseNumber(CaseData caseData) {
+        return isNull(caseData.getFamilyManCaseNumber()) ? "" : caseData.getFamilyManCaseNumber() + ",";
     }
 
-    private String getHearingBooking(CaseData data) {
+    protected String getLeadRespondentsName(CaseData caseData) {
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/base/StandardDirectionOrderContent.java:17]: I feel like this method can be inlined and the callers instead just use the method that this one is using

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/LocalAuthorityEmailContentProvider.java [hunk 10]
```diff
@@ -5,26 +5,34 @@
...
 
 @Service
 @RequiredArgsConstructor(onConstructor = @__(@Autowired))
 public class LocalAuthorityEmailContentProvider extends StandardDirectionOrderContent {
     private final LocalAuthorityNameLookupConfiguration config;
 
-    public Map<String, Object> buildLocalAuthorityStandardDirectionOrderIssuedNotification(CaseData caseData) {
-        return super.getSDOPersonalisationBuilder(caseData)
-            .put("title", config.getLocalAuthorityName(caseData.getCaseLocalAuthority()))
+    public SDONotifyData buildLocalAuthorityStandardDirectionOrderIssuedNotification(CaseData caseData) {
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/LocalAuthorityEmailContentProvider.java:20]:
```suggestion
    public SDONotifyData buildStandardDirectionOrderIssuedNotification(CaseData caseData) {
```
we already know it is for the LA from the content provider

**Review by lewisbirks — COMMENTED** [review]
Looks good, few changes suggested. Would be happier with multiple eyes looking over it, there is a lot to cover and it is very repetitive so I am almost certain that I will have overlooked something.

Also, please look at removing the method in the NotifyData interface. Looks like it is only used in one other place so should be simple enough to remove.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/notify/allocatedjudge/SharedAllocatedJudgeTemplate.java (hunk 2 above)
Comment (aliveni) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/notify/allocatedjudge/SharedAllocatedJudgeTemplate.java:10]: no longer required, removed it

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/notify/sendtogatekeeper/NotifyGatekeeperTemplate.java (hunk 6 above)
Comment (aliveni) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/notify/sendtogatekeeper/NotifyGatekeeperTemplate.java:13]: I tried it but it's causing a code smell suggesting to access the static builder. sonarcloud reported the same earlier

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/notify/submittedcase/SubmitCaseCafcassTemplate.java (hunk 7 above)
Comment (aliveni) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/notify/submittedcase/SubmitCaseCafcassTemplate.java:12]: as discussed, using the `toBuilder` method to make the field private

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/base/AbstractEmailContentProvider.java (hunk 8 above)
Comment (aliveni) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/base/AbstractEmailContentProvider.java]: removed it

**Comment by aliveni** [comment]
> Looks good, few changes suggested. Would be happier with multiple eyes looking over it, there is a lot to cover and it is very repetitive so I am almost certain that I will have overlooked something.
> 
> Also, please look at removing the method in the NotifyData interface. Looks like it is only used in one other place so should be simple enough to remove.

Removed the method from the NotifyData interface

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/notify/hearing/UpcomingHearingNotifyData.java [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import lombok.Builder;
+import lombok.Data;
+import uk.gov.hmcts.reform.fpl.model.notify.NotifyData;
+
+@Builder
+@Data
+public class UpcomingHearingNotifyData implements NotifyData {
+    @JsonProperty("hearing_date")
```
Comment (samuelzerozero) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/notify/hearing/UpcomingHearingNotifyData.java:11]: did we know/test if that this serialisation works?

**Review by samuelzerozero — APPROVED** [review]

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/notify/hearing/UpcomingHearingNotifyData.java (hunk 11 above)
Comment (aliveni) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/notify/hearing/UpcomingHearingNotifyData.java:11]: added a test in `UpcomingHearingsContentProviderTest`

**Review by samuelzerozero — APPROVED** [review]

**Pull request merged**